compiler.py

```diff
@@ -0,0 +1,637 @@
+                return str(node.value)
+            elif isinstance(node.value, str):
+                return '"' + ''.join(c for c in node.value if c >= ' ' and c != '"') + '"' #scrub quotes and control characters from string
+            else:
+                raise CompilerError(ERR_COMPLEX_VALUE, node)
+
+        elif isinstance(node, ast.Name):
+            assert isinstance(node.ctx, ast.Load)
+            # TODO should this raise a CompilerError like most other errors do?
+            return node.id
+            # TODO figure out what this does???
+
+        elif isinstance(node, ast.Call):
+            if isinstance(node.func, ast.Attribute):
+                if node.func.value.id == 'Env':
+                    return self.env_as_value(node.func)
+                else:
+                    raise CompilerError(ERR_COMPLEX_VALUE, node)
+            # TODO  It looks like this is the place to handle object attributes like vault1.copper and Unit.x
+
+            fn = self._functions.get(node.func.id)
+            if fn is None:
+                raise CompilerError(ERR_NO_DEF, node)
+
+            if len(node.args) != fn.argc:
+                raise CompilerError(ERR_ARGC_MISMATCH, node)
+
+            for arg in node.args:
+                val = self.as_value(arg)
+                self._ins.append(f'write {val} cell1 {REG_STACK}')
+                self._ins.append(f'op add {REG_STACK} {REG_STACK} 1')
+            # TODO  probably better to set the values straight into the relevant variables, rather than slowly piping through stack
+
+            # `REG_STACK` is not updated because it's immediately read by the function.
+            # If it was, the `op add` followed by the `op sub` would be redundant.
+            self._ins.append(f'write @counter cell1 {REG_STACK}')
+            self._ins.append(f'jump {fn.start} always')
+            return REG_RET
+            # TODO  probably better to pass the returned value via a dummy variable rather than piping through stack
+
+        else:
+            raise CompilerError(ERR_COMPLEX_VALUE, node)
+            # TODO This may be where we want code that handles complex expressions, like 1+2*3
+
+    def env_as_value(self, node):
+        """This is used to process expressions like Env.width, which compile to mlog globals like @mapw """
+        var = ENV_TO_VAR.get(node.attr)
+        if var is None:
+            raise CompilerError(ERR_UNSUPPORTED_SYSCALL, node)
+        return var
+
+    def generate_masm(self):
+        """This will be called after the full list self._ins of mlog instructions is generated, concatenating
+           them together into a single string of mlog assembly (masm) code."""
+        if len(self._ins) + 1 > MAX_INSTRUCTIONS:
+            raise CompilerError(ERR_TOO_LONG, ast.Module(lineno=0, col_offset=0))
+
+        return '\n'.join(self._ins) + '\nend\n'
+        #TODO it's not at all clear that you really want an explicit 'end' here!  My understanding is that

```

Comment:
The `end` is here because `if` conditions emit a jump to the next instruction, and with no `end`, that jump is out of bounds, which I believe doesn't really work? Needs testing.

Programs should also not rely on variables/values from previous invocations, because that's not how Python program works. To persist data one uses files, or in Mindustry, we would use memory cells.

Reply:
Yeah, I'd made a TODO to test what happens if you jump out of bounds too.  If it is a problem, it's trivially easy with my new label system to redirect such out-of-bounds labels back to line 0, which would be better than forcing variable wipes on users who don't want them.

I think we may have a slight difference in design philosophy/goals here.  I'd like a convenient way of writing mlog programs that will take advantage of strengths of mlog, and work around its more annoying/tedious aspects.  So, e.g., I want Pyndustric to give me access to mlog `jump` commands to arbitrary destinations, and the fact that Python itself doesn't offer such jumps counts absolutely zero in my mind against offering this in Pyndustric.  Similarly, mlog has a strict limit on program length, but effectively wraps my whole program in an implicit `while True` loop, without counting that against my instruction count.  The fact that Python doesn't default to having such a loop surround its programs again counts absolutely zero in my mind against allowing my Pyndustric programs to take advantage of it. 

(A bigger concern in this ballpark is that updating the code on a mindustry processor doesn't always reset its state or even its @counter, which can lead to some pretty frustrating bugs.  If one of these options made such frustrations less likely, I'd give that positive weight.  But I'm not sure this choice makes any difference to that, and I'm also not sure how often people will update running processors with new Pyndustric compilations, and I think that that sort of code-update might be one that does reset state and @counter anyway, so this whole concern may be moot.)

Reply:
> effectively wraps my whole program in an implicit `while True` loop, without counting that against my instruction count

This is a good point.

> updating the code on a mindustry processor doesn't always reset its state or even its `@counter`

This was fixed after I nagged Anuken about it (https://github.com/Anuken/Mindustry-Suggestions/issues/1702 + linked issue), which I'm grateful for.